moon/utils/server/sv_powerups_manager.moon

```diff
@@ -28,3 +28,46 @@ PowerupManager =
 
         if existingPowerup
             existingPowerup\Refresh!
+
+    getRandomPowerup: (tier) ->
+        CFCBasePowerup = CFCPowerups["base_cfc_powerup"]
+
+        sumOfWeights = CFCBasePowerup.powerupTotalWeights[tier] - 1
+        randomIndex = math.random 0, sumOfWeights
+
+        for name, powerup in pairs CFCBasePowerup.powerupList
+            if randomIndex < powerup.powerupWeights[tier]
+                return powerup
+
+            randomIndex -= powerup.powerupWeights[tier]
+
+    getShuffledSpawnLocations: (tier) ->
+        spawnLocations = table.Copy CFCPowerups.spawnLocations[tier]
+
+        for i = #spawnLocations, 2, -1
+            j = math.random i
+            spawnLocations[i], spawnLocations[j] = spawnLocations[j], spawnLocations[i]
+
+        spawnLocations
+
+    getRandomSpawnLocations: (tier) ->
+        allSpawnLocations = PowerupManager.getShuffledSpawnLocations tier
+        spawnCount = math.floor #allSpawnLocations / (tier + 1)
+
+        spawnLocations = {}
+
+        for i = 1, spawnCount
+            table.insert spawnLocations, allSpawnLocations[i]
+
+        spawnLocations
+
+    spawnRandomPowerups: ->
+        for tier = 1, 4 do
+            spawnLocations = PowerupManager.getRandomSpawnLocations tier
+
+            for _, location in ipairs spawnLocations
+                powerupClass = PowerupManager.getRandomPowerup!.powerupID
+
+                powerup = ents.Create powerupClass
+                powerup\SetPos location
+                powerup\Spawn!
```

Comment:
I think this part should be moved into its own file in the `utils/server/` folder.

Maybe `moon/utils/server/sv_powerups_spawner.moon` or something?